### PR TITLE
Us015 edit role api

### PR DESCRIPTION
**Pull request by BlazejRaj**
there is some code regarding Capability that i created because of mistakes in specification. I hope its ok :)

### CODE REVIEW

File: src/main/java/com/kainos/ea/dao/RolesDao.java [hunk 1]
```diff
@@ -43,6 +43,66 @@ public List<JobRole> getAllRoles(Connection c) throws SQLException {
...
+                    resultSet.getString("kainos_job_title"),
+                    resultSet.getString("job_specification"),
+                    resultSet.getString("job_spec_link")
+            );
+            return jobRoleXL;
+        }
+        return null;
+    }
+
+    //TODO: what to return with update ?
```
Comment (TomaszLaski) [inline src/main/java/com/kainos/ea/dao/RolesDao.java]: remove comments

Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/dao/RolesDao.java]: done

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 2]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
 
+    @GET
+    @Path("/job-roles/{id}")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getJobRoleById(@PathParam("id") int id){
+        try {
+            return Response.status(HttpStatus.OK_200).entity(rolesService.getRoleById(id)).build();
+        } catch (Exception | DatabaseConnectionException  e){
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
+        } catch (JobRoleDoesNotExistException e){
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java:68]: this needs to go above the other catch as the other catch catches generic exception class

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 3]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+            return Response.status(HttpStatus.NOT_FOUND_404, e.getMessage()).build();
+        }
+    }
+
+    @PUT
+    @Path("/job-roles")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response updateJobRole(JobRoleXL jobRoleXL){
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: this should use the same model as the create job role endpoint

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 4]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+            return Response.status(HttpStatus.OK_200).entity(rolesService.getRoleById(id)).build();
+        } catch (Exception | DatabaseConnectionException  e){
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
+        } catch (JobRoleDoesNotExistException e){
+            return Response.status(HttpStatus.NOT_FOUND_404, e.getMessage()).build();
+        }
+    }
+
+    @PUT
+    @Path("/job-roles")
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: URL should have the id of the role you want to edit

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 5]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+                try {
+                    int id = rolesService.updateJobRole(jobRoleXL);
+                    return Response.status(HttpStatus.OK_200).entity(id).build();
+                } catch (Exception | DatabaseConnectionException e) {
+                    return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500).build();
+                }
+            } else {
+                return Response.status(HttpStatus.BAD_REQUEST_400).build();
+            }
+        } catch (LinkSyntaxException e) {
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: I'm not sure what this exception is supposed to be? maybe its just poorly names?

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 6]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getBandById(@PathParam("id") int id){
+        try {
+            return Response.status(HttpStatus.OK_200).entity(bandsService.getBandById(id)).build();
+        } catch (Exception | DatabaseConnectionException  e){
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
+        } catch (BandDoesNotExistException e){
+            return Response.status(HttpStatus.NOT_FOUND_404, e.getMessage()).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: not sure why this endpoint is needed for your story?

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 7]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getFamilyById(@PathParam("id") int id){
+        try {
+            return Response.status(HttpStatus.OK_200).entity(familiesService.getFamilyById(id)).build();
+        } catch (Exception | DatabaseConnectionException  e){
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
+        } catch (FamilyDoesNotExistException e){
+            return Response.status(HttpStatus.NOT_FOUND_404, e.getMessage()).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: not sure why this endpoint is needed for your story?

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 8]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response updateJobFamily(JobFamily jobFamily){
+        try {
+            int id = familiesService.updateJobFamily(jobFamily);
+            return Response.status(HttpStatus.OK_200).entity(id).build();
+        } catch (Exception | DatabaseConnectionException e) {
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: not sure why this endpoint is needed for your story?

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 9]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+    public Response getCapabilityById(@PathParam("id") int id){
+        try {
+            return Response.status(HttpStatus.OK_200).entity(capabilitiesService.getCapabilityById(id)).build();
+        } catch (Exception | DatabaseConnectionException  e){
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
+        } catch (CapabilityDoesNotExistException e){
+            return Response.status(HttpStatus.NOT_FOUND_404, e.getMessage()).build();
+        }
+    }
+
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: not sure why this endpoint is needed for your story?

File: src/main/java/com/kainos/ea/controller/WebService.java [hunk 10]
```diff
@@ -38,6 +61,125 @@ public Response getJobRoles()
...
+    @GET
+    @Path("/capabilities")
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response getCapabilities(){
+        try{
+            return Response.status(HttpStatus.OK_200).entity(capabilitiesService.getAllCapabilities()).build();
+        } catch (Exception | DatabaseConnectionException e) {
+            return Response.status(HttpStatus.INTERNAL_SERVER_ERROR_500, e.getMessage()).build();
+        }
+    }
```
Comment (shaunganley) [inline src/main/java/com/kainos/ea/controller/WebService.java]: not sure why this endpoint is needed for your story?

**Review by shaunganley — DISMISSED** [review]

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 2 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java:68]: done

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 6 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: removed

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 7 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: removed

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 8 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java:157]: its needed to populate dropdown list

Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: removed

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 9 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: removed

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 10 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: removed

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 5 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: changed

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 4 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: done

**Review by TomaszLaski — APPROVED** [review]

File: src/main/java/com/kainos/ea/controller/WebService.java (hunk 3 above)
Comment (BlazejRaj) [reply src/main/java/com/kainos/ea/controller/WebService.java]: changing JobRole to JobRoleResponse, JobRoleXL to JobRoleRequest

File: src/test/java/com/kainos/ea/service/RolesServiceTest.java [hunk 11]
```diff
@@ -53,6 +44,38 @@ void getAllRoles_shouldThrowException_whenDaoThrowsException() throws SQLExcepti
         assertThrows(SQLException.class, () -> rolesService.getAllRoles());
     }
 
+    @Test
+    void getRoleById_shouldReturnJobRoleXLWhenDaoReturnsJobRoleXL() throws SQLException, DatabaseConnectionException, IOException, JobRoleDoesNotExistException {
```
Comment (boydstepnew) [inline src/test/java/com/kainos/ea/service/RolesServiceTest.java]: This needs renamed, no longer XL

Comment (BlazejRaj) [reply src/test/java/com/kainos/ea/service/RolesServiceTest.java]: done

**Review by boydstepnew — APPROVED** [review]

**Pull request merged**